pkg/collector/upgrade/upgrade.go

```diff
@@ -73,23 +68,36 @@ func (u VersionUpgrade) ManagedInstances(ctx context.Context) error {
 			u.Recorder.Event(&original, "Error", "Upgrade", msg)
 			continue
 		}
-
 		if !reflect.DeepEqual(upgraded, list.Items[i]) {
 			// the resource update overrides the status, so, keep it so that we can reset it later
 			st := upgraded.Status
-			patch := client.MergeFrom(&original)
-			if err := u.Client.Patch(ctx, &upgraded, patch); err != nil {
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				col := v1beta1.OpenTelemetryCollector{}
+				err = u.Client.Get(ctx, client.ObjectKeyFromObject(&original), &col)
+				if err != nil {
+					return err
+				}
+				var upgradedErr error
+				upgraded, upgradedErr = u.ManagedInstance(ctx, col)
+				if upgradedErr != nil {
+					return upgradedErr
+				}
+				if err := u.Client.Update(ctx, &upgraded); err != nil {
+					return err
+				}
+				return nil
+			})
```

Comment:
i found a bug in the upgrade that was preventing the 105 from being applied. The patch never applied the removal of the unify env feature gate which meant that the 107 release was causing anything that had that gate to fail rollout. When i moved to update, i started getting the good ol' object has changed bug, so i wrapped this in the retry on conflict method from [this PR](https://github.com/open-telemetry/opentelemetry-operator/pull/2383)

```
{"level":"ERROR","timestamp":"2024-08-23T20:36:50Z","logger":"collector-upgrade","message":"failed to apply changes to instance","name":"opentelemetry-kube-stack-cluster-stats","namespace":"default","error":"Operation cannot be fulfilled on opentelemetrycollectors.opentelemetry.io \"opentelemetry-kube-stack-cluster-stats\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/open-telemetry/opentelemetry-operator/pkg/collector/upgrade.VersionUpgrade.ManagedInstances\n\t/Users/jacob.aronoff/workspace/playground/opentelemetry-operator/pkg/collector/upgrade/upgrade.go:88\nmain.addDependencies.func1\n\t/Users/jacob.aronoff/workspace/playground/opentelemetry-operator/main.go:458\nsigs.k8s.io/controller-runtime/pkg/manager.RunnableFunc.Start\n\t/Users/jacob.aronoff/workspace/playground/opentelemetry-operator/vendor/sigs.k8s.io/controller-runtime/pkg/manager/manager.go:307\nsigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1\n\t/Users/jacob.aronoff/workspace/playground/opentelemetry-operator/vendor/sigs.k8s.io/controller-runtime/pkg/manager/runnable_group.go:226"}
```

If anyone can explain why patch was working to _add_ the new args but fail to _remove_ the unwanted arg I would love to know.